python/ray/serve/batching.py

```diff
@@ -141,20 +150,46 @@ async def _handle_batches(self, func):
             try:
                 # Method call.
                 if self_arg is not None:
-                    results = await func(self_arg, *args, **kwargs)
+                    func_future_or_generator = func(self_arg, *args, **kwargs)
                 # Normal function call.
                 else:
-                    results = await func(*args, **kwargs)
-
-                if len(results) != len(batch):
-                    raise RayServeException(
-                        "Batched function doesn't preserve batch size. "
-                        f"The input list has length {len(batch)} but the "
-                        f"returned list has length {len(results)}."
-                    )
-
-                for i, result in enumerate(results):
-                    futures[i].set_result(result)
+                    func_future_or_generator = func(*args, **kwargs)
+                if isasyncgenfunction(func):
```

Comment:
This function is getting really hard to follow now, can you extract the generator- and non-generator handling into separate functions?

Reply:
I refactored the function and pulled some logic out to helper functions.